relay/shfu_grpc/client.go

```diff
@@ -0,0 +1,212 @@
+package shfu_grpc
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/ibc-go/v8/modules/core/exported"
+	"github.com/datachainlab/lcp-go/relay/shfu_logger"
+	"github.com/datachainlab/lcp-go/relay/shfu_storage"
+	"github.com/google/uuid"
+	"github.com/hyperledger-labs/yui-relayer/core"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// GetLatestFinalizedHeader retrieves the latest finalized header from gRPC server
+func GetLatestFinalizedHeader(ctx context.Context, grpcAddress string, chainID string, counterpartyChainID string, codec codec.ProtoCodecMarshaler) (core.Header, error) {
+	logger := shfu_logger.GetSHFULogger(ctx)
+	requestID := uuid.Must(uuid.NewV7()).String()
+
+	logger.InfoContext(ctx, "GetLatestFinalizedHeader request started",
+		"request_id", requestID,
+		"grpc_address", grpcAddress,
+		"chain_id", chainID,
+		"counterparty_chain_id", counterpartyChainID)
+
+	if grpcAddress == "" {
+		logger.ErrorContext(ctx, "GetLatestFinalizedHeader request failed - no gRPC address", fmt.Errorf("SHFU gRPC address not provided"),
+			"request_id", requestID,
+			"chain_id", chainID,
+			"counterparty_chain_id", counterpartyChainID)
+		return nil, fmt.Errorf("SHFU gRPC address not provided")
+	}
+
+	// Connect to SHFU gRPC server
+	conn, err := grpc.DialContext(ctx, grpcAddress,
```

Comment:
`DialContext` is deprecated:

```suggestion
	conn, err := grpc.NewClient(grpcAddress,
```

Reply:
fixed